sklearn/datasets/samples_generator.py

```diff
@@ -28,8 +27,7 @@ def _generate_hypercube(samples, dimensions, rng):
     if dimensions > 30:
         return np.hstack([_generate_hypercube(samples, dimensions - 30, rng),
                           _generate_hypercube(samples, 30, rng)])
-    out = astype(sample_without_replacement(2 ** dimensions, samples,
-                                            random_state=rng),
+    out = astype(random_state.permutation(2 ** dimensions)[:samples],
```

Comment:
So should it go back to calling "sample_without_replacement" and then have auto determine whether to use permutation or not, or should it just make that determination directly in this one case and only call sample_without_replacement in the edge cases?